File: sdk/python/examples/sdk_basic.py
#!/usr/bin/env python3
"""Basic SDK usage example — Agent.create(), session, and tool execution.

Demonstrates:
  - Loading config from `.a3s/config.hcl`
  - Creating agents with Agent.create()
  - Creating workspace-bound sessions
  - Direct tool calls via session.tool(name, args) for all built-in tools
  - Convenience methods: bash(), read_file(), glob(), grep()

No LLM calls — runs entirely offline.

Usage:
    # Build the native module first
    cd crates/code/sdk/python && maturin develop --release

    # Run the example
    python examples/sdk_basic.py

    # Or with a custom config path
    A3S_CONFIG=/path/to/config.hcl python examples/sdk_basic.py
"""

import os
import sys
import tempfile
import time

from a3s_code import Agent


def repo_root():
    """Resolve the repo root: this file is at crates/code/sdk/python/examples/"""
    return os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))


def resolve_config():
    """Resolve config path: $A3S_CONFIG > repo-root/.a3s/config.hcl"""
    env = os.environ.get("A3S_CONFIG")
    if env:
        return env
    return os.path.join(repo_root(), ".a3s", "config.hcl")


def main():
    config_path = resolve_config()
    print("=== A3S Code Python SDK Basic Example ===\n")
    print(f"Config: {config_path}\n")

    # -- 1. Agent.create() ---------------------------------------------------
    print("--- Agent.create() ---")
    agent = Agent.create(config_path)
    print("[ok] Agent created\n")

    # -- 2. Create session ----------------------------------------------------
    workspace = tempfile.mkdtemp(prefix="a3s-py-")
    session = agent.session(workspace)
    print(f"[ok] Session bound to {workspace}\n")

    # -- 3. Tool: bash --------------------------------------------------------
    print("--- Tool: bash ---")
    output = session.bash("echo 'Hello from Python SDK'")
    assert "Hello from Python SDK" in output, f"unexpected: {output}"
    print(f"  output: {output.strip()}")

    # -- 4. Tool: write + read ------------------------------------------------
    print("\n--- Tool: write + read ---")
    file_path = os.path.join(workspace, "demo.txt")
    session.tool("write", {
        "file_path": file_path,
        "content": "Hello from Python SDK!\nLine 2\nLine 3",
    })
    print(f"  wrote: {file_path}")

    content = session.read_file(file_path)
    print(f"  read back ({len(content.splitlines())} lines):")
    for line in content.splitlines()[:3]:
        print(f"    {line}")

    # -- 5. Tool: edit --------------------------------------------------------
    print("\n--- session.tool('edit') ---")
    edit_result = session.tool("edit", {
        "file_path": file_path,
        "old_string": "Line 2",
        "new_string": "Line 2 (edited)",
    })
    print(f"  exit_code: {edit_result.exit_code}")
    assert edit_result.exit_code == 0, f"edit failed: {edit_result.output}"
    edited = open(os.path.join(workspace, "demo.txt")).read()
    assert "Line 2 (edited)" in edited, "edit not applied"
    print("  verified: file contains 'Line 2 (edited)'")

    # -- 6. Tool: patch -------------------------------------------------------
    print("\n--- session.tool('patch') ---")
    patch_result = session.tool("patch", {
        "file_path": file_path,
        "diff": "@@ -2,2 +2,2 @@\n-Line 2 (edited)\n+Line 2 (patched)\n Line 3",
    })
    print(f"  exit_code: {patch_result.exit_code}")
    if patch_result.exit_code == 0:
        patched = open(os.path.join(workspace, "demo.txt")).read()
        assert "Line 2 (patched)" in patched, "patch not applied"
        print("  verified: file contains 'Line 2 (patched)'")
    else:
        print(f"  patch output: {patch_result.output.strip()}")

    # -- 7. Tool: read --------------------------------------------------------
    print("\n--- session.tool('read') ---")
    read_result = session.tool("read", {
        "file_path": file_path,
        "offset": 0,
        "limit": 5,
    })
    print(f"  exit_code: {read_result.exit_code}")
    assert read_result.exit_code == 0, f"read failed: {read_result.output}"
    print("  content:")
    for line in read_result.output.splitlines()[:3]:
        print(f"    {line}")

    # -- 8. Tool: ls ----------------------------------------------------------
    print("\n--- session.tool('ls') ---")
    # Create some test files
    with open(os.path.join(workspace, "alpha.rs"), "w") as f:
        f.write("fn main() {}")
    with open(os.path.join(workspace, "beta.rs"), "w") as f:
        f.write("fn test() {}")
    with open(os.path.join(workspace, "gamma.txt"), "w") as f:
        f.write("text file")
    os.makedirs(os.path.join(workspace, "subdir"), exist_ok=True)

    ls_result = session.tool("ls", {"path": workspace})
    print(f"  exit_code: {ls_result.exit_code}")
    assert ls_result.exit_code == 0, f"ls failed: {ls_result.output}"
    assert "alpha.rs" in ls_result.output, "ls should list alpha.rs"
    assert "subdir" in ls_result.output, "ls should list subdir"
    print("  entries:")
    for line in ls_result.output.splitlines()[:6]:
        print(f"    {line}")

    # -- 9. Tool: glob --------------------------------------------------------
    print("\n--- session.tool('glob') ---")
    glob_result = session.tool("glob", {"pattern": "*.rs"})
    print(f"  exit_code: {glob_result.exit_code}")
    assert glob_result.exit_code == 0, f"glob failed: {glob_result.output}"
    assert "alpha.rs" in glob_result.output, "glob should find alpha.rs"
    print(f"  matches: {glob_result.output.strip()}")

    # -- 10. Tool: grep -------------------------------------------------------
    print("\n--- session.tool('grep') ---")
    grep_result = session.tool("grep", {"pattern": "fn ", "glob": "*.rs"})
    print(f"  exit_code: {grep_result.exit_code}")
    assert grep_result.exit_code == 0, f"grep failed: {grep_result.output}"
    assert "fn " in grep_result.output, "grep should find 'fn '"
    print("  output:")
    for line in grep_result.output.splitlines()[:5]:
        print(f"    {line}")

    # -- 11. Tool: bash (direct) ----------------------------------------------
    print("\n--- session.tool('bash') ---")
    bash_result = session.tool("bash", {
        "command": "echo 'direct tool call' && date +%Y",
        "timeout": 5000,
    })
    print(f"  exit_code: {bash_result.exit_code}")
    assert bash_result.exit_code == 0, f"bash failed: {bash_result.output}"
    assert "direct tool call" in bash_result.output
    print(f"  output: {bash_result.output.strip()}")

    # -- 12. Tool: web_fetch --------------------------------------------------
    print("\n--- session.tool('web_fetch') ---")
    fetch_result = session.tool("web_fetch", {
        "url": "https://httpbin.org/get",
        "format": "text",
        "timeout": 15,
    })
    print(f"  exit_code: {fetch_result.exit_code}")
    assert fetch_result.exit_code == 0, f"web_fetch failed: {fetch_result.output}"
    print("  response (first 5 lines):")
    for line in fetch_result.output.splitlines()[:5]:
        print(f"    {line}")

    # -- 13. Tool: web_search -------------------------------------------------
    print("\n--- session.tool('web_search') ---")
    search_result = session.tool("web_search", {
        "query": "Rust programming language",
        "engines": "ddg,wiki",
        "limit": 3,
        "timeout": 15,
        "format": "text",
    })
    print(f"  exit_code: {search_result.exit_code}")
    if search_result.exit_code == 0:
        print("  results (first 5 lines):")
        for line in search_result.output.splitlines()[:5]:
            print(f"    {line}")
    else:
        print(f"  [soft-fail] web_search error: {search_result.output.splitlines()[0] if search_result.output else '(empty)'}")

    # -- 14. Tool: cron (full lifecycle) --------------------------------------
    print("\n--- session.tool('cron') -- full lifecycle ---")
    root = repo_root()
    cron_session = agent.session(root)
    print(f"  workspace: {root} (cron data -> .a3s/cron/)")

    # parse
    print("\n  [parse] 'every 5 minutes'")
    parse_result = cron_session.tool("cron", {"action": "parse", "input": "every 5 minutes"})
    print(f"    exit_code: {parse_result.exit_code}")
    assert parse_result.exit_code == 0, f"cron parse failed: {parse_result.output}"
    assert "*/5" in parse_result.output
    print(f"    output: {parse_result.output.strip()}")

    # add
    job_name = f"py-sdk-test-{int(time.time())}"
    print(f"\n  [add] job '{job_name}' schedule='*/10 * * * *'")
    add_result = cron_session.tool("cron", {
        "action": "add",
        "name": job_name,
        "schedule": "*/10 * * * *",
        "command": "echo 'python sdk cron test'",
    })
    print(f"    exit_code: {add_result.exit_code}")
    assert add_result.exit_code == 0, f"cron add failed: {add_result.output}"

    # Extract job ID
    job_id = None
    for line in add_result.output.splitlines():
        stripped = line.strip()
        if stripped.startswith("ID: "):
            job_id = stripped[4:]
            break
    assert job_id, f"could not extract job ID from: {add_result.output}"
    print(f"    job_id: {job_id}")

    # list
    print("\n  [list]")
    list_result = cron_session.tool("cron", {"action": "list"})
    assert list_result.exit_code == 0
    assert job_name in list_result.output
    print(f"    found '{job_name}' in list")

    # get
    print(f"\n  [get] id={job_id}")
    get_result = cron_session.tool("cron", {"action": "get", "id": job_id})
    assert get_result.exit_code == 0
    assert job_name in get_result.output
    print(f"    output: {get_result.output.splitlines()[0]}")

    # pause
    print(f"\n  [pause] id={job_id}")
    pause_result = cron_session.tool("cron", {"action": "pause", "id": job_id})
    assert pause_result.exit_code == 0
    print(f"    output: {pause_result.output.strip()}")

    # resume
    print(f"\n  [resume] id={job_id}")
    resume_result = cron_session.tool("cron", {"action": "resume", "id": job_id})
    assert resume_result.exit_code == 0
    print(f"    output: {resume_result.output.strip()}")

    # run
    print(f"\n  [run] id={job_id}")
    run_result = cron_session.tool("cron", {"action": "run", "id": job_id})
    assert run_result.exit_code == 0
    print(f"    output: {run_result.output.splitlines()[0]}")

    # history
    print(f"\n  [history] id={job_id}")
    history_result = cron_session.tool("cron", {"action": "history", "id": job_id, "limit": 5})
    assert history_result.exit_code == 0
    print(f"    output: {history_result.output.splitlines()[0]}")

    # remove
    print(f"\n  [remove] id={job_id}")
    remove_result = cron_session.tool("cron", {"action": "remove", "id": job_id})
    assert remove_result.exit_code == 0
    print(f"    output: {remove_result.output.strip()}")

    # verify removal
    list_after = cron_session.tool("cron", {"action": "list"})
    assert job_name not in list_after.output
    print("    verified: job no longer in list")

    # -- 15. Unknown tool (error case) ----------------------------------------
    print("\n--- session.tool('nonexistent') ---")
    unknown_result = session.tool("nonexistent", {})
    print(f"  exit_code: {unknown_result.exit_code}")
    assert unknown_result.exit_code == 1, "unknown tool should fail"
    assert "Unknown tool" in unknown_result.output
    print(f"  output: {unknown_result.output.strip()}")

    # -- 16. Convenience methods ----------------------------------------------
    print("\n--- Convenience methods ---")
    glob_conv = session.glob("*.rs")
    print(f"  session.glob('*.rs'): {glob_conv}")

    grep_conv = session.grep("fn ")
    print(f"  session.grep('fn '): {len(grep_conv.splitlines())} lines")

    # -- Done -----------------------------------------------------------------
    print("\n=== All checks passed ===")


if __name__ == "__main__":
    main()

File: sdk/node/examples/sdk_basic.mjs
#!/usr/bin/env node
/**
 * Basic SDK usage example — Agent.create(), session, and tool execution.
 *
 * Demonstrates:
 *   - Loading config from `.a3s/config.hcl`
 *   - Creating agents with Agent.create()
 *   - Creating workspace-bound sessions
 *   - Direct tool calls via session.tool(name, args) for all built-in tools
 *   - Convenience methods: bash(), readFile(), glob(), grep()
 *
 * No LLM calls — runs entirely offline.
 *
 * Usage:
 *     # Build the native module first
 *     cd crates/code/sdk/node && npm run build
 *
 *     # Run the example
 *     node examples/sdk_basic.mjs
 *
 *     # Or with a custom config path
 *     A3S_CONFIG=/path/to/config.hcl node examples/sdk_basic.mjs
 */

import { Agent } from '../index.js';
import { mkdtempSync, writeFileSync, readFileSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import { tmpdir } from 'os';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function repoRoot() {
  return join(__dirname, '..', '..', '..', '..', '..');
}

function resolveConfig() {
  if (process.env.A3S_CONFIG) return process.env.A3S_CONFIG;
  return join(repoRoot(), '.a3s', 'config.hcl');
}

function assert(condition, message) {
  if (!condition) throw new Error(`Assertion failed: ${message}`);
}

async function main() {
  const configPath = resolveConfig();
  console.log('=== A3S Code Node.js SDK Basic Example ===\n');
  console.log(`Config: ${configPath}\n`);

  // -- 1. Agent.create() ---------------------------------------------------
  console.log('--- Agent.create() ---');
  const agent = await Agent.create(configPath);
  console.log('[ok] Agent created\n');

  // -- 2. Create session ----------------------------------------------------
  const workspace = mkdtempSync(join(tmpdir(), 'a3s-node-'));
  const session = agent.session(workspace);
  console.log(`[ok] Session bound to ${workspace}\n`);

  // -- 3. Tool: bash --------------------------------------------------------
  console.log('--- Tool: bash ---');
  const output = await session.bash("echo 'Hello from Node.js SDK'");
  assert(output.includes('Hello from Node.js SDK'), `unexpected: ${output}`);
  console.log(`  output: ${output.trim()}`);

  // -- 4. Tool: write + read ------------------------------------------------
  console.log('\n--- Tool: write + read ---');
  const filePath = join(workspace, 'demo.txt');
  await session.tool('write', {
    file_path: filePath,
    content: 'Hello from Node.js SDK!\nLine 2\nLine 3',
  });
  console.log(`  wrote: ${filePath}`);

  const content = await session.readFile(filePath);
  const lines = content.split('\n');
  console.log(`  read back (${lines.length} lines):`);
  lines.slice(0, 3).forEach(line => console.log(`    ${line}`));

  // -- 5. Tool: edit --------------------------------------------------------
  console.log('\n--- session.tool("edit") ---');
  const editResult = await session.tool('edit', {
    file_path: filePath,
    old_string: 'Line 2',
    new_string: 'Line 2 (edited)',
  });
  console.log(`  exit_code: ${editResult.exitCode}`);
  assert(editResult.exitCode === 0, `edit failed: ${editResult.output}`);
  const edited = readFileSync(join(workspace, 'demo.txt'), 'utf-8');
  assert(edited.includes('Line 2 (edited)'), 'edit not applied');
  console.log("  verified: file contains 'Line 2 (edited)'");

  // -- 6. Tool: patch -------------------------------------------------------
  console.log('\n--- session.tool("patch") ---');
  const patchResult = await session.tool('patch', {
    file_path: filePath,
    diff: "@@ -2,2 +2,2 @@\n-Line 2 (edited)\n+Line 2 (patched)\n Line 3",
  });
  console.log(`  exit_code: ${patchResult.exitCode}`);
  if (patchResult.exitCode === 0) {
    const patched = readFileSync(join(workspace, 'demo.txt'), 'utf-8');
    assert(patched.includes('Line 2 (patched)'), 'patch not applied');
    console.log("  verified: file contains 'Line 2 (patched)'");
  } else {
    console.log(`  patch output: ${patchResult.output.trim()}`);
  }

  // -- 7. Tool: read --------------------------------------------------------
  console.log('\n--- session.tool("read") ---');
  const readResult = await session.tool('read', {
    file_path: filePath,
    offset: 0,
    limit: 5,
  });
  console.log(`  exit_code: ${readResult.exitCode}`);
  assert(readResult.exitCode === 0, `read failed: ${readResult.output}`);
  console.log('  content:');
  readResult.output.split('\n').slice(0, 3).forEach(line => console.log(`    ${line}`));

  // -- 8. Tool: ls ----------------------------------------------------------
  console.log('\n--- session.tool("ls") ---');
  writeFileSync(join(workspace, 'alpha.rs'), 'fn main() {}');
  writeFileSync(join(workspace, 'beta.rs'), 'fn test() {}');
  writeFileSync(join(workspace, 'gamma.txt'), 'text file');
  mkdirSync(join(workspace, 'subdir'), { recursive: true });

  const lsResult = await session.tool('ls', { path: workspace });
  console.log(`  exit_code: ${lsResult.exitCode}`);
  assert(lsResult.exitCode === 0, `ls failed: ${lsResult.output}`);
  assert(lsResult.output.includes('alpha.rs'), 'ls should list alpha.rs');
  assert(lsResult.output.includes('subdir'), 'ls should list subdir');
  console.log('  entries:');
  lsResult.output.split('\n').slice(0, 6).forEach(line => console.log(`    ${line}`));

  // -- 9. Tool: glob --------------------------------------------------------
  console.log('\n--- session.tool("glob") ---');
  const globResult = await session.tool('glob', { pattern: '*.rs' });
  console.log(`  exit_code: ${globResult.exitCode}`);
  assert(globResult.exitCode === 0, `glob failed: ${globResult.output}`);
  assert(globResult.output.includes('alpha.rs'), 'glob should find alpha.rs');
  console.log(`  matches: ${globResult.output.trim()}`);

  // -- 10. Tool: grep -------------------------------------------------------
  console.log('\n--- session.tool("grep") ---');
  const grepResult = await session.tool('grep', { pattern: 'fn ', glob: '*.rs' });
  console.log(`  exit_code: ${grepResult.exitCode}`);
  assert(grepResult.exitCode === 0, `grep failed: ${grepResult.output}`);
  assert(grepResult.output.includes('fn '), "grep should find 'fn '");
  console.log('  output:');
  grepResult.output.split('\n').slice(0, 5).forEach(line => console.log(`    ${line}`));

  // -- 11. Tool: bash (direct) ----------------------------------------------
  console.log('\n--- session.tool("bash") ---');
  const bashResult = await session.tool('bash', {
    command: "echo 'direct tool call' && date +%Y",
    timeout: 5000,
  });
  console.log(`  exit_code: ${bashResult.exitCode}`);
  assert(bashResult.exitCode === 0, `bash failed: ${bashResult.output}`);
  assert(bashResult.output.includes('direct tool call'));
  console.log(`  output: ${bashResult.output.trim()}`);

  // -- 12. Tool: web_fetch --------------------------------------------------
  console.log('\n--- session.tool("web_fetch") ---');
  const fetchResult = await session.tool('web_fetch', {
    url: 'https://httpbin.org/get',
    format: 'text',
    timeout: 15,
  });
  console.log(`  exit_code: ${fetchResult.exitCode}`);
  assert(fetchResult.exitCode === 0, `web_fetch failed: ${fetchResult.output}`);
  console.log('  response (first 5 lines):');
  fetchResult.output.split('\n').slice(0, 5).forEach(line => console.log(`    ${line}`));

  // -- 13. Tool: web_search -------------------------------------------------
  console.log('\n--- session.tool("web_search") ---');
  const searchResult = await session.tool('web_search', {
    query: 'Rust programming language',
    engines: 'ddg,wiki',
    limit: 3,
    timeout: 15,
    format: 'text',
  });
  console.log(`  exit_code: ${searchResult.exitCode}`);
  if (searchResult.exitCode === 0) {
    console.log('  results (first 5 lines):');
    searchResult.output.split('\n').slice(0, 5).forEach(line => console.log(`    ${line}`));
  } else {
    const firstLine = searchResult.output.split('\n')[0] || '(empty)';
    console.log(`  [soft-fail] web_search error: ${firstLine}`);
  }

  // -- 14. Tool: cron (full lifecycle) --------------------------------------
  console.log('\n--- session.tool("cron") -- full lifecycle ---');
  const root = repoRoot();
  const cronSession = agent.session(root);
  console.log(`  workspace: ${root} (cron data -> .a3s/cron/)`);

  // parse
  console.log("\n  [parse] 'every 5 minutes'");
  const parseResult = await cronSession.tool('cron', { action: 'parse', input: 'every 5 minutes' });
  console.log(`    exit_code: ${parseResult.exitCode}`);
  assert(parseResult.exitCode === 0, `cron parse failed: ${parseResult.output}`);
  assert(parseResult.output.includes('*/5'));
  console.log(`    output: ${parseResult.output.trim()}`);

  // add
  const jobName = `node-sdk-test-${Date.now()}`;
  console.log(`\n  [add] job '${jobName}' schedule='*/10 * * * *'`);
  const addResult = await cronSession.tool('cron', {
    action: 'add',
    name: jobName,
    schedule: '*/10 * * * *',
    command: "echo 'node sdk cron test'",
  });
  console.log(`    exit_code: ${addResult.exitCode}`);
  assert(addResult.exitCode === 0, `cron add failed: ${addResult.output}`);

  // Extract job ID
  let jobId = null;
  for (const line of addResult.output.split('\n')) {
    const trimmed = line.trim();
    if (trimmed.startsWith('ID: ')) {
      jobId = trimmed.slice(4);
      break;
    }
  }
  assert(jobId, `could not extract job ID from: ${addResult.output}`);
  console.log(`    job_id: ${jobId}`);

  // list
  console.log('\n  [list]');
  const listResult = await cronSession.tool('cron', { action: 'list' });
  assert(listResult.exitCode === 0);
  assert(listResult.output.includes(jobName));
  console.log(`    found '${jobName}' in list`);

  // get
  console.log(`\n  [get] id=${jobId}`);
  const getResult = await cronSession.tool('cron', { action: 'get', id: jobId });
  assert(getResult.exitCode === 0);
  assert(getResult.output.includes(jobName));
  console.log(`    output: ${getResult.output.split('\n')[0]}`);

  // pause
  console.log(`\n  [pause] id=${jobId}`);
  const pauseResult = await cronSession.tool('cron', { action: 'pause', id: jobId });
  assert(pauseResult.exitCode === 0);
  console.log(`    output: ${pauseResult.output.trim()}`);

  // resume
  console.log(`\n  [resume] id=${jobId}`);
  const resumeResult = await cronSession.tool('cron', { action: 'resume', id: jobId });
  assert(resumeResult.exitCode === 0);
  console.log(`    output: ${resumeResult.output.trim()}`);

  // run
  console.log(`\n  [run] id=${jobId}`);
  const runResult = await cronSession.tool('cron', { action: 'run', id: jobId });
  assert(runResult.exitCode === 0);
  console.log(`    output: ${runResult.output.split('\n')[0]}`);

  // history
  console.log(`\n  [history] id=${jobId}`);
  const historyResult = await cronSession.tool('cron', { action: 'history', id: jobId, limit: 5 });
  assert(historyResult.exitCode === 0);
  console.log(`    output: ${historyResult.output.split('\n')[0]}`);

  // remove
  console.log(`\n  [remove] id=${jobId}`);
  const removeResult = await cronSession.tool('cron', { action: 'remove', id: jobId });
  assert(removeResult.exitCode === 0);
  console.log(`    output: ${removeResult.output.trim()}`);

  // verify removal
  const listAfter = await cronSession.tool('cron', { action: 'list' });
  assert(!listAfter.output.includes(jobName));
  console.log('    verified: job no longer in list');

  // -- 15. Unknown tool (error case) ----------------------------------------
  console.log('\n--- session.tool("nonexistent") ---');
  const unknownResult = await session.tool('nonexistent', {});
  console.log(`  exit_code: ${unknownResult.exitCode}`);
  assert(unknownResult.exitCode === 1, 'unknown tool should fail');
  assert(unknownResult.output.includes('Unknown tool'));
  console.log(`  output: ${unknownResult.output.trim()}`);

  // -- 16. Convenience methods ----------------------------------------------
  console.log('\n--- Convenience methods ---');
  const globConv = await session.glob('*.rs');
  console.log(`  session.glob('*.rs'): ${JSON.stringify(globConv)}`);

  const grepConv = await session.grep('fn ');
  console.log(`  session.grep('fn '): ${grepConv.split('\n').length} lines`);

  // -- Done -----------------------------------------------------------------
  console.log('\n=== All checks passed ===');
}

main().catch(err => {
  console.error('FAILED:', err);
  process.exit(1);
});
